### ISSUE TITLE
Feature request: Add OR option to conditions

### ISSUE THREAD

**Issue by costa-alex** [strautomator/web#11]
Some of the conditions could have an OR option to allow more flexible automations.

Something like:
Condition: GPS device / app name has TrainerRoad OR Zwift

**Comment by sevitz** [comment]
+1 to that request.

Would be really useful to go 
("If starting location is HOME" OR "if end location is home") 
AND
activity is walk
then
Set GEAR to walking shoes)

Etc

**Comment by mitchgibbstek** [comment]
Just adding another vote for this. I can see how it would complicate things, but it would be handy. I've got a bunch of workouts from our Tonal smart gym that come through as weight training, but they're really other modalities. Unfortunately, the only thing I can do is create separate automations to recategorize each one by name. It would be handy if I could do something like "If name is x or y or z then change type to yoga". That would drop it to just a few automations.

I did have the insane thought that maybe if there were an option to allow RegEx, the highly motivated few could use it for more complicated scenarios without complicating things for the majority. However, the idea that RegEx simplifies anything makes my head spin :-)

**Comment by igoramadas** [comment]
Well that's 3 people now, so... I'll investigate how much work is needed to add a flag to change the operator from AND to OR on a given automation. Or to allow a single condition to have many fields to compare to (so OR on a single condition instead of the whole set of conditions).

It won't happen this week thou, as I'm currently on a cycling holidays and should be back only late next week.

@mitchgibbstek regex would open a whole can of worms. Way too many things to consider security and performance wise. I know it's technically possible, but that would be a very, very niche use case so keeping it out of scope for now.

**Comment by sevitz** [comment]
Ah thats great. Thanks for considering.

Have a great holiday. 🚲

**Comment by mitchgibbstek** [comment]
Wow! My vote may have pushed it over the threshold? See kids, every vote matters. 🤣 Enjoy your holiday and thanks both for looking at adding the feature and for creating such a useful tool. My friends thank you are spared being spammed by my workouts ;-)

**Comment by Knagie** [comment]
I only just now discovered the Conditions area was AND and not OR. I have 8 different buildings I sometimes need to travel to, either at the start of the day or travel between them during the day. I added all of these as "End location is X" and was wondering why that didn't work. Is there an obvious way to know conditions is AND and not OR, without trying like I did?

I would love to have my commutes all covered in 1 automation instead of the 8 I would need to create now just for my morning commutes. As well as (yea I'm picky) the 8 for the commute home which might start at any of the 8 locations. With an OR option that could all be set in 1 automation instead of 16 duplicates with just a different start/end address :)

And just setting it to "End at home" wouldn't work as I also have other bike trips that end back at home. 

So make that 6 people now looking for the feature? 😄

**Comment by igoramadas** [comment]
The problem with making automation recipes with ANDs and ORs, as stated before, is that it's quite clear for programmers and people with a technical background to understand it, but extremely confusing for everyone else. People will wonder what is the correct order, will an OR negate a previous AND, etc... And considering that I am a moron regarding UI and UX skills, I'm pretty sure the end result would suck.

I am thinking about making all action fields multi-valued (like it already happens for sport types and week days, for example). That way, you could say "if location end is X or Y or Z". Or "distance is A or B". This is easy to do, and pretty simple to explain / lay out on the UI. I'll get this done and publish it to the beta website for testing sometime next week (too busy at work these days, sorry).

@sevitz in your example, can't you simply use "Passes on location" instead of start / end? Unless your house is mid-way on some walk activities and you want to exclude those :-)

**Comment by sevitz** [comment]
Ah didn't see that was an option. Useful. Can use that for a bunch.

Think the modal might also solve a different bunch, which should reduce the need for a pure OR operator in a lot of cases (although as a (former) programmer I can still see a AND/OR option being useful)

**Comment by igoramadas** [comment]
Logical operators now available for testing: https://beta.strautomator.com/

PS: the beta environment is restricted to PRO users, and at the moment the UI is still not working with Google Maps (for example, when fetching a location for an automation condition). Maps will be enabled sometime this week.

Still need to do some polishing.

![Preview](https://user-images.githubusercontent.com/1576674/221687741-ca6028b5-2428-4161-ab6b-450337a367d1.png)

**Comment by Knagie** [comment]
That was really quick! Thanks for this, I'll mess about with it once I'm back home.

**Comment by mitchgibbstek** [comment]
I'm testing it out, creating an automation with two triggers conditions of different types (word exists in Name or Description) with OR with actions to change the sport type, mute, and stop further processing. It won't let me save it. It acts like I haven't clicked the button. Not sure if that's intentional. I've tried a couple of times and tried simplifying the criteria and actions but I get the same result.

The Logical Operators make sense to me and seem like a good compromise, but I could imagine them potentialy confusing some people. The visual, after you've selected one or both of the options, makes it very clear to me what it's going to do though,

**Comment by igoramadas** [comment]
@mitchgibbstek can you please press F12 (or open your browser's dev tool) and check if there are any Javascript or network errors on the console? I just tried a few random automations now using my account, but had no issues, so I need to know how to replicate this.

**Comment by mitchgibbstek** [comment]
@igoramadas  It looks like it's CORS. 

```
Access to XMLHttpRequest at 'https://strautomator.com/api/users/10969429/recipes' from origin 'https://beta.strautomator.com' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.
489a54e.js:1 
        
       AutomationEdit.save o
(anonymous) @ 489a54e.js:1
strautomator.com/api/users/10969429/recipes:1 

```

**Comment by igoramadas** [comment]
Can you try again? I have changed the build process to properly set the axios URLs now.

**Comment by mitchgibbstek** [comment]
@igoramadas Yep! It's working now. 

I created an automation with multiple criteria of the same type (Name has). I ran it against several Strava activities that, I think, it should match. It worked for two where the names matched exactly. 

For another, I think it should do a partial match - it's looking for "Day of Focus" and the activity's name is "Day Of Focus: Wave" but it didn't pick it up when I fed the URL to the automation. If I misunderstand how it works and that it has to be exact matches, "is" or "equals" might be clearer than "has". If partial matches are okay, then "contains" feels more consistent with how this sort of thing shows in other applications. Just my opinion though.

I also noticed different behavior when choosing Conditions or Actions allowing me to type part of the entry. It works much better than how it's working for me in production.

Thanks!

**Comment by igoramadas** [comment]
@mitchgibbstek thanks for the feedback. I will consider rephrasing some things to make sure everything is crystal clear.

You are correct: with so many conditions and actions now, I switched to an autocomplete box where users can type what they want. Production is still using the old style (simple select box).

**Comment by igoramadas** [comment]
Feature is going live today. Beta testing was quite limited (only 7 users), so I don't think all possible use cases were tested, but I'll keep the existing instances running in parallel so if anything goes wrong I can quickly switch back. Otherwise, feedback is always appreciated.

**Comment by igoramadas** [comment]
https://github.com/strautomator/web/releases/tag/v23.97.11555

**Comment by Knagie** [comment]
It works great for me, though it's so far only limited to my above use case. It does safe 15 automations so that's great. 
Probably also saves computing time as I can let it stop after marking it through my commute options instead of going through every automation.

UI was pretty clear to me, the AND and OR options change at the top as well if you change both so it clearly displays how it should work. And testing it with a manual activity is great to see if it works as intended.
